src/pip/_internal/utils/misc.py

```diff
@@ -966,6 +966,21 @@ def split_auth_from_netloc(netloc):
     return netloc, user_pass
 
 
+def split_auth_netloc_from_url(original_url):
```

Comment:
Since this method so closely parallels the structure of `_transform_url()` below, I think it would help if it directly preceded it. (The unneeded newlines can also be removed.)

Reply:
It seemed to fit well coming immediately after `split_auth_from_netloc`, while `_transform_url` is very much a helper function for those that follow it.

I just had a look at enhancing `_transform_url` to be able to return extra extracted information, but everything ended up far more complex than simply duplicating the calls to urlsplit/urlunsplit throughout.

Reply:
What I wanted to draw attention to is that both functions have the same "split" followed by "unsplit" flow (and I think they're the only ones), so they should be kept in sync or be candidates for merging. For example, it seems weird that one uses `urlparse()` / `urlunparse()` while the other uses `urlsplit()` / `urlunsplit()` (should they be the same?). I doubt that was intentional, but I could be wrong.

Reply:
No, it wasn't intentional. Looks like `download.py` uses `urlparse` while `misc.py` uses `urlsplit` and I didn't update it in the copy paste.

In any case, I came up with a fairly neat way to unify them all under `_transform_url`, so that's coming next.